src/dali_executor/io_buffer.h

```diff
@@ -0,0 +1,135 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2021 NVIDIA CORPORATION
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef TRITONDALIBACKEND_IO_BUFFER_H
+#define TRITONDALIBACKEND_IO_BUFFER_H
+
+#include "src/dali_executor/io_descriptor.h"
+#include "src/dali_executor/utils/dali.h"
+
+namespace triton { namespace backend { namespace dali {
+
+
+void CopyMem(device_type_t dst_dev, void *dst, device_type_t src_dev, const void *src, size_t size,
+             cudaStream_t stream = 0);
+
+class IOBufferI {
+ public:
+  /**
+   * Reserve a chunk of `size` of bytes and return a pointer
+   * to the beginning of the chunk.
+   */
+  virtual uint8_t *Reserve(size_t size) = 0;
+
+  /**
+   * Cancel all reservations. No memory is deallocated.
+   */
+  virtual void Clear() = 0;
+
+  /**
+   * Allocate `size` bytes of memory.
+   * If the buffer's capacity is greater or equal to size, this function is a no-op.
+   */
+  virtual void Allocate(size_t size) = 0;
```

Comment:
I would swap names Allocate and Reserve. As Reserve usually doesn't provide any useful memory, while allocate does. In this case it is the opposite.

Reply:
done